src/main/java/org/passay/LengthComplexityRule.java

```diff
@@ -0,0 +1,413 @@
+package org.passay;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Rule for determining if a password contains the desired complexity for a certain length. In order to meet the
+ * criteria of this rule, passwords must meet all the supplied rules for a given password length.
+ *
+ * @author  Middleware Services
+ */
+public class LengthComplexityRule implements Rule
+{
+
+  /** Error code for insufficient complexity. */
+  public static final String ERROR_CODE = "INSUFFICIENT_COMPLEXITY";
+
+  /** Rules to apply when checking a password. */
+  private final Map<Interval, List<Rule>> rules = new HashMap<>();
+
+  /** Whether to report the details of each complexity rule failure. */
+  private boolean reportRuleFailures = true;
+
+
+  /**
+   * Adds the rules to invoke for the supplied interval.
+   *
+   * @param  interval  of integers that the supplied rules apply to
+   * @param  l  list of rules
+   *
+   * @throws  IllegalArgumentException  if the supplied rules are empty or null or if interval is invalid or intersects
+   *                                    with an existing interval
+   */
+  public void addRules(final String interval, final List<Rule> l)
+  {
+    if (l == null || l.isEmpty()) {
+      throw new IllegalArgumentException("Rules cannot be empty or null");
+    }
+    final Interval i = new Interval(interval);
+    for (Interval existingInterval : rules.keySet()) {
+      if (existingInterval.intersects(i)) {
+        throw new IllegalArgumentException("Interval " + i + " intersects existing interval " + existingInterval);
+      }
+    }
+    rules.put(i, l);
+  }
+
+
+  /**
+   * Adds the rules to invoke for the supplied interval.
+   *
+   * @param  interval  of integers that the supplied rules apply to
+   * @param  r  list of rules
+   */
+  public void addRules(final String interval, final Rule... r)
+  {
+    addRules(interval, r != null ? Arrays.asList(r) : null);
+  }
+
+
+  /**
+   * Returns whether to add the rule result detail for each rule that fails to validate to the rule result.
+   *
+   * @return  whether to add rule result details
+   */
+  public boolean getReportRuleFailures()
+  {
+    return reportRuleFailures;
+  }
+
+
+  /**
+   * Sets whether to add the rule result detail for each rule that fails to validate to the rule result.
+   *
+   * @param  b  whether to add rule result details
+   */
+  public void setReportRuleFailures(final boolean b)
+  {
+    reportRuleFailures = b;
+  }
+
+
+  @Override
+  public RuleResult validate(final PasswordData passwordData)
+  {
+    final List<Rule> rulesByLength = getRulesByLength(passwordData.getPassword().length());
+    if (rulesByLength == null) {
+      throw new IllegalStateException(
```

Comment:
We discussed allowing this case to succeed to allow other rules to take effect when an input is outside a configured range. How do you feel about a flag to permit that behavior?